module/VuFind/tests/integration-tests/src/VuFindTest/Config/ConfigWritingTest.php

```diff
@@ -0,0 +1,204 @@
+        }
+    }
+
+    /**
+     * Upgrade test provider.
+     *
+     * @return array[]
+     */
+    public static function upgradeTestProvider(): array
+    {
+        return [
+            'generic file handler' => [
+                'generic-file',
+                'config',
+            ],
+            'ini handler' => [
+                'ini',
+                'config',
+            ],
+            'dir handler' => [
+                'dir',
+                'baseDir',
+            ],
+        ];
+    }
+
+    /**
+     * Test writing.
+     *
+     * @param string $fixture    Fixture
+     * @param string $configName Config name
+     *
+     * @return void
+     *
+     * @dataProvider upgradeTestProvider
+     */
+    public function testWriting(string $fixture, string $configName): void
+    {
+        $container = $this->getContainerWithConfigRelatedServices(
+            baseDir: $this->getFixtureDir() . 'configs/write/' . $fixture,
+            baseSubDir: ''
+        );
+        $pathResolver = $container->get(PathResolver::class);
+        $configManager = $container->get(ConfigManager::class);
+
+        $baseDirPath = $pathResolver->getBaseConfigDirPath();
+        $baseConfigLocation = $pathResolver->getMatchingConfigLocation($baseDirPath, $configName);
+
+        $destinationLocation = clone $baseConfigLocation;
+        $destinationLocation->setBasePath($this->localDirPath);
+
+        $config = $configManager->loadConfigFromLocation($baseConfigLocation);
+        $configManager->writeConfig($destinationLocation, $config, $baseConfigLocation);
+
+        $this->assertDirsEqual($baseDirPath, $this->localDirPath);
+    }
+
+    /**
+     * Assert that two configuration dirs are equal.
+     *
+     * @param string $expected Expected directory
+     * @param string $actual   Actual directory
+     *
+     * @return void
+     */
+    protected function assertDirsEqual(string $expected, string $actual): void
+    {
+        $this->assertDirectoryExists($expected);
+        $this->assertDirectoryExists($actual);
+
+        $expectedContent = scandir($expected);
+        $actualContent = scandir($actual);
+        $this->assertEquals($expectedContent, $actualContent);
+
+        foreach ($expectedContent as $item) {
+            if ($item == '.' || $item == '..') {
+                continue;
+            }
+            if (is_dir($expected . '/' . $item)) {
+                $this->assertDirsEqual($expected . '/' . $item, $actual . '/' . $item);
+            } else {
+                $expectedFileContent = $this->readFile($expected . '/' . $item);
+                $actualFileContent = $this->readFile($actual . '/' . $item);
+                $this->assertEquals($expectedFileContent, $actualFileContent);
+            }
+        }
+    }
+
+    /**
+     * Read file and clean up whitespaces.
+     *
+     * @param string $path File path
+     *
+     * @return string
+     */
+    protected function readFile(string $path): string
```

Comment:
Would it make sense to give this method a more descriptive name?

```suggestion
    protected function readFileAndNormalizeWhitespace(string $path): string
```

...it also feels like this method and assertDirsEqual are generically useful enough that maybe we should consider introducing a trait to make them reusable. Of course, we could wait unitl there's a second use case before doing so.

Reply:
I feel like right now with the normalization of white spaces this is specialized for this test. But off course, we can create a trait if we want to do something similar somewhere else.